components/mpas-ocean/src/shared/mpas_ocn_vertical_remap.F

```diff
@@ -0,0 +1,436 @@
+      integer :: cell1, cell2, iCell, jCell, iEdge, iTrac, k, kTop, kBot
+
+      type (mpas_pool_type), pointer :: verticalMeshPool
+      type (mpas_pool_type), pointer :: statePool
+      type (mpas_pool_type), pointer :: tracersPool
+      type (mpas_pool_iterator_type) :: groupItr
+
+      type(rmap_work) :: work
+      type(rcon_ends), dimension(:), allocatable :: bcUpper, bcLower
+
+      real (kind=RKIND), dimension(:, :, :), allocatable :: dataNow, dataNew
+      real (kind=RKIND), dimension(:, :), allocatable :: &
+         layerThicknessNew,            & ! layerThickness (new target grid)
+         layerThickEdge,               & ! layerThickness at edges (lagrangian grid)
+         layerThickEdgeNew,            & ! layerThickness at edges (new target grid)
+         heightCellNow, heightEdgeNow, & ! depth at layer interfaces (lagrangian grid)
+         heightCellNew, heightEdgeNew    ! depth at layer interfaces (new target grid)
+
+      real (kind=RKIND), dimension(:, :), pointer :: layerThickness, restingThickness
+      real (kind=RKIND), dimension(:, :), pointer :: normalVelocity
+      real (kind=RKIND), dimension(:, :), pointer :: highFreqThickness,lowFreqDivergence
+      real (kind=RKIND), dimension(:, :, :), pointer :: tracersGroup
+
+      err = 0
+
+      itimestepLastRemap = itimestepLastRemap + 1
+      if (itimestepLastRemap < config_vert_remap_interval) return
+      itimestepLastRemap = 0
+
+      ! Remapping currently only supports one block for the whole domain rather
+      ! than sub-blocks
+      call mpas_pool_get_subpool(block % structs, 'verticalMesh', verticalMeshPool)
+      call mpas_pool_get_subpool(block % structs, 'state', statePool)
+
+      call mpas_pool_get_subpool(statePool, 'tracers', tracersPool)
+
+      call mpas_pool_get_array(verticalMeshPool, 'restingThickness', restingThickness)
+
+      call mpas_pool_get_array(statePool, 'layerThickness', layerThickness, 2)
+      call mpas_pool_get_array(statePool, 'normalVelocity', normalVelocity, 2)
+      if (config_use_freq_filtered_thickness) then
+         call mpas_pool_get_array(statePool, 'highFreqThickness', highFreqThickness, 2)
+         call mpas_pool_get_array(statePool, 'lowFreqDivergence', lowFreqDivergence, 2)
+      end if
+
+      nCells = nCellsAll
+      nEdges = nEdgesAll
+
+      ! SETUP VERTICAL GRIDS:
+
+      allocate(heightCellNow(nVertLevels + 1, nCells))
+      allocate(heightCellNew(nVertLevels + 1, nCells))
+      allocate(heightEdgeNow(nVertLevels + 1, nEdges))
+      allocate(heightEdgeNew(nVertLevels + 1, nEdges))
+      allocate(layerThicknessNew(nVertLevels, nCells+1))
+      allocate(layerThickEdge(nVertLevels, nEdges))
+      allocate(layerThickEdgeNew(nVertLevels, nEdges))
+
+      ! Compute new layer thicknesses based on vertical coordinate choice
+      ! Currently, zstar coordinate is hard-coded
+      call ocn_vert_regrid(restingThickness, layerThickness, &
+                               layerThicknessNew, err)
```

Comment:
I don't think values of `layerThicknessNew` that aren't valid (the dummy cell at `nCells+1` and the cells above `minLevelCell` or below `maxLevelCell` ever get assigned a value.  I think that's likely to be a problem because the whole array gets copied into `layerThickness`.  I think it needs to be assigned a value that is the same as the missing value for `layerThickness` before this call.  This will mean you need to get the field for `layerThickness` and not just the array so you have access to the missing value attribute.

Reply:
The same may apply to other fields that you allocate here but I didn't see clear cases where those fields get copied in full.  If `minLevelCell` or `maxLevelCell` could change between remapping, it would be important to make sure all fields are filled with appropriate missing values above and below the new min and max if they are allowed to change.

Reply:
@xylar Thanks for clarifying. Here https://github.com/E3SM-Ocean-Discussion/E3SM/pull/6/commits/8827635428203a86d4b992b7aea5b78b5478d232 I went the route of allocating `layerThicknessNew` to the size of `layerThickness` and assigning all values to `0`, the current missing value, in `ocn_vertical_regrid` before filling it with the regridded values so that the current `minLevelCell` and `maxLevelCell` limits are used.